### PR TITLE
fix(security): require explicit approval for shell redirection/background in command auto-approve

### PR DESCRIPTION
**Pull request by 0xMink**
Refs #11095

## Summary

- Command auto-approval could approve an allowlisted prefix while shell operators (redirection/background) alter execution semantics under `shell: true`. For example, auto-approving `git show` would also auto-approve `git show > out.txt`, allowing writes to an arbitrary file.
- Commands containing file redirection (`<`, `>`, `>>`, `<<`, etc.) now require explicit approval (`ask_user`).
  - Safe fd-to-fd duplication is excluded (for example `2>&1`, `>&2`, `<&3`, `0<&4`).
- Commands containing a standalone background operator (`&`) now require explicit approval.
- fd-to-fd stripping now uses a **token-boundary** lookahead `(?=$|\s|[;&|()<>])` to avoid false negatives such as `>&2file` and `<&3in`, where the redirection target is a word and therefore represents file redirection, not fd-to-fd duplication.

## Test plan

- `containsShellFileRedirection()` — 21 cases: detects output/input/append/here-doc/stderr/mixed redirections (9), excludes safe fd-to-fd like `2>&1`, `>&2`, `<&3`, `0<&4` (5), token-boundary edge cases — `>&2file`, `<&3in`, `2>&1&&`, `2>&1>out.txt`, `0<&4|` (5), general sanity checks (2)
- `containsBackgroundOperator()` — 8 cases: detects standalone `&`, excludes `&&`, `&>`, `>&`, `<&`
- `getCommandDecision()` — 15 integration cases: redirection forces `ask_user`, fd-to-fd preserves `auto_approve`, background forces `ask_user`, token-boundary edge cases, denylist regression
- Existing regression coverage for `containsDangerousSubstitution`, `findLongestPrefixMatch`, `getSingleCommandDecision` — 11 cases
- Total: 55 tests, all passing

### CODE REVIEW

**Comment by ghost** [comment]
> **Rooviewer**  Clock &nbsp;&nbsp;[See task](https://app.roocode.com/cloud-jobs/f69zsFuG?utm_source=github-comment&utm_medium=link&utm_campaign=github.pr.review.sync)

All issues from previous reviews have been addressed. Quote-aware operator detection via `stripQuotedSegments` is correctly implemented. No new issues found.

- [x] `containsShellFileRedirection` strips output fd-to-fd (`>&N`) but not input fd-to-fd (`<&N`), producing false positives for commands like `cmd <&3`. Very low severity.

> <details>
> <summary>Previous reviews</summary>
> 
> - 0d2a949: [Review #1](https://app.roocode.com/cloud-jobs/YGyCPOKB?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 2ab6380: [Review #2](https://app.roocode.com/cloud-jobs/Qi1bdAnf?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - cf5084f: [Review #3](https://app.roocode.com/cloud-jobs/H1qLy7Zx?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 4c99f4f: [Review #4](https://app.roocode.com/cloud-jobs/y5qYWIzR?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> </details>
<sub>Mention @roomote in a comment to request specific changes to this pull request or [fix all unresolved issues](https://app.roocode.com/fix-issues?repo=RooCodeInc/Roo-Code&prNumber=11367&commentId=3875576044).</sub>

File: src/core/auto-approval/commands.ts [hunk 1]
```diff
@@ -64,6 +64,45 @@ export function containsDangerousSubstitution(source: string): boolean {
...
+ * This strips safe fd-to-fd redirections (2>&1, >&2) and then checks for any
+ * remaining < or > characters, which indicate file redirection.
+ *
+ * @param command - The command string to analyze
+ * @returns true if file redirection operators are detected
+ */
+export function containsShellFileRedirection(command: string): boolean {
+	// Strip fd-to-fd redirections (2>&1, >&2, 1>&2) which don't write to files.
+	// Requires at least one digit after & to avoid stripping >&file patterns.
+	const stripped = command.replace(/\d*>&\d+/g, "")
```
Comment (ghost) [inline src/core/auto-approval/commands.ts]: Minor: the stripping regex handles output fd-to-fd (`>&N`) but not input fd-to-fd (`<&N`). Shell commands like `cmd <&3` or `cmd 0<&3` are safe fd-to-fd duplications that would produce false positives here because `<` survives the strip and triggers the check. Very low severity since `<&N` is rare in auto-approved commands and the result is just an unnecessary prompt, but for completeness the strip could cover both directions:

```suggestion
	const stripped = command.replace(/\d*[<>]&\d+/g, "")
```

<sub>[Fix it with Roo Code](https://app.roocode.com/fix-issue/?repo=RooCodeInc/Roo-Code&prNumber=11367&fixId=3d700b9660) or mention @roomote and request a fix.</sub>

**Comment by 0xMink** [comment]
Addressed the review feedback:

- **Fixed false negative:** Added token-boundary lookahead to fd-to-fd stripping so `>&2file` and `<&3in` are correctly treated as file redirection (not fd-to-fd). The boundary `(?=$|\s|[;&|()<>])` ensures stripping only when the fd target is purely numeric and token-delimited.
- **Added input fd-to-fd support:** `<&N` and `0<&N` patterns are now stripped alongside `>&N`, and excluded from background operator detection.
- **Kept correct behavior:** `2>&1&&` still strips; `0<&4|` still strips; `>&2` at end-of-string still strips.
- **Consistent stripping:** Both `containsShellFileRedirection()` and the sub-command normalization in `getCommandDecision()` use the same boundary-aware regex.
- **Tests:** 55/55 passing (up from 44). Added regression tests for `>&2file`, `<&3in`, operator-adjacent fd-to-fd, and input fd duplication at both unit and integration levels.

**Comment by 0xMink** [comment]
@roomote Could you please re-review? I addressed your <&N note, and also tightened fd-to-fd stripping with a token-boundary lookahead to avoid false negatives like >&2file / <&3in. Tests expanded from 44 to 55 and I force-pushed the update. Thanks.

**Comment by ghost** [comment]
> **Fixaroo**  Clock &nbsp;&nbsp;[See task](https://app.roocode.com/cloud-jobs/qEr3g3vm?utm_source=github-comment&utm_medium=link&utm_campaign=github.pr.comment.respond)

No outstanding issues to fix. The reviewer has confirmed all previous issues are addressed, and all 55 tests pass locally.

**Comment by 0xMink** [comment]
Rebased onto main, resolved add/add in commands.spec.ts with #11365 / #11382, and made redirection/background detection quote-aware to avoid false positives in quoted payloads (e.g. node -e "const f=(a)=>a"). All CI checks are green.

@roomote please re-review.

**Comment by ghost** [comment]
> **Fixaroo**  Clock &nbsp;&nbsp;[See task](https://app.roocode.com/cloud-jobs/trVOCpoh?utm_source=github-comment&utm_medium=link&utm_campaign=github.pr.comment.respond)

Verified the latest state of the PR. The reviewer confirmed all previous issues are resolved and no new issues were found. All 79 tests pass locally. No code changes needed.

**Comment by hannesrudolph** [comment]
Closing: PR lacks clear evidence of manual testing.

**Comment by robertjmcintyre** [comment]
@0xMink - I think that this is a fairly important issue, especially as people get more and more comfortable with giving Roo increasing autonomy. Is it possible to get this reactivated and committed?
